packages/vite-plugin-express/src/index.ts

```diff
@@ -25,75 +56,145 @@ const getPluginConfig = async (server: ViteDevServer): Promise<VitePluginExpress
+
+    const modules = Array.from(server.moduleGraph.getModulesByFile(file) ?? []);
+    if (modules.length === 0) {
+      // This file isn't part of the module graph. No need to do anything.
+      return false;
+    }
+
+    if (modules.length !== 1) {
+      throw new Error(`Unexpected module graph state: ${modules.length} modules found for ${file}`);
+    }
+
+    // If the file is directly or indirectly imported by a module that matches
+    // the full restart paths, then we need to do a full restart.
+    const importers = recursivelyGetImporters(modules[0]);
+    const importerPaths = Array.from(importers)
+      .map((m) => m.file)
+      .filter((p): p is string => p !== null);
+    return importerPaths.some((p) => fullRestartPathMatchers.some((matcher) => matcher(p)));
+  }
+
   // We'll manually react to changes in two cases:
   //
-  // 1. If the file matches any of the extra watch paths, we'll immediately restart
-  //    the server. These files aren't part of Vite's module graph, so we can't rely
-  //    on any automatic reloading. This also means we need to restart the entire
-  //    server, as we can't just reload the module.
+  // 1. If the file matches any of the full paths, we'll immediately restart
+  //    the server. These paths may not be part of the Vite module graph (e.g.
+  //    SQL files), or they may be part of the module graph but may contain state
+  //    that can't be hot-reloaded (e.g. database connection pools or worker pools).
   //
   // 2. If we're configured to watch file changes, we'll reload the app (root) module.
   //    This gives us a head start over waiting for the next request to trigger a reload.
   server.watcher.on('change', (file) => {
-    if (extraWatchPathMatchers.some((matcher) => matcher(file))) {
+    if (needsFullRestart(path.resolve(server.config.root, file))) {
+      logger.info(`Change to ${file} requires a full restart`, { timestamp: true });
       debouncedRestart();
     } else if (config.watchFileChanges) {
       debouncedLoadApp();
     }
   });
 
-  return async function (req: IncomingMessage, res: ServerResponse): Promise<void> {
-    const app = await _loadApp(config);
-    if (!app) {
-      return;
-    }
+  // Hook into the environment's `close` method so we can do cleanup when the
+  // server is closed. Crucially, this allows us to hook into the `k+Enter` key
+  // combo in the Vite dev server to gracefully shut down the server so that it
+  // can be restarted cleanly.
+  //
+  // This uses a hardcoded `close` export from the app module.
+  const originalClose = server.environments.ssr.close.bind(server.environments.ssr);
+  server.environments.ssr.close = async () => {
+    await originalClose();
 
-    app.use((err: unknown, _req: typeof req, _res: typeof res, next: Connect.NextFunction) => {
-      if (err instanceof Error) {
-        server.ssrFixStacktrace(err);
-      }
+    if (typeof _mostRecentAppModule?.close === 'function') {
```

Comment:
We may want to consider a config option for `close`, since there is a config option for `exportName`.

Reply:
Sounds good, I'd rather just get this merged.